### PR TITLE
Ananjan/regularizer changes

### PR DESCRIPTION
**Pull request by ananjan-nandi-9**
This PR introduces several new features for the tree regularizer, including:
1. Soft split decisions at tree nodes using gumbel softmax
2. Difference-based SCI score computation. A span is represented by the difference in lm embeddings of its last token and the token before its first token.
3. Phrase sampling for efficient computation.
4. Regularization based on root-level split decision only. Given a sentence, split it into two context-free spans.
5. Regularization guided by gold parses for SimPL.
6. Shikhar's idea for a linear time complexity regularizer. 

This PR also contains a quick hack to get tree regularization scores over a given dataset for any dumped model.

### CODE REVIEW

File: data_utils/ds_addmult_mod10_helpers.py [hunk 1]
```diff
@@ -105,7 +106,9 @@ def build_dataset_addmult_mod10(
         # sample unique subtrees
         ds_uniques = set(dataset.unique('tree_sig'))
         n_subtrees = int(len(ds_uniques) * hold_out_p_subtrees)
-        held_out_subtrees = set(random.sample(ds_uniques, n_subtrees))
+        # This is not reproducible
+        # held_out_subtrees = set(random.sample(ds_uniques, n_subtrees))
+        held_out_subtrees = set(list(ds_uniques)[:n_subtrees])
```
Comment (mcgrath-k) [inline data_utils/ds_addmult_mod10_helpers.py]: This is fine with me, but won't this require running a full epoch to get a proper understanding of the performance of network if the distribution of generated samples is uneven? Might want to check with Derek or highlight this one so that everyone is onboard.

File: train_transformers.py [hunk 2]
```diff
@@ -167,51 +168,7 @@ def get_datasets_and_vocab(args, language_model: bool):
...
-    """
-    Main function for language modeling tasks.
-
-    Args:
-        args: Command line arguments.
-    """
-    language_model = args.lm
-    out_vocab = CharVocabulary(chars=set('0123456789'))
-
-    datasets, in_vocab = get_datasets_and_vocab(args, language_model)
```
Comment (mcgrath-k) [inline train_transformers.py:214]: Seems like github glitched due to the relocation of line `171`. Just to check, did you run this after merging them locally? The file itself seems fine.

File: training_utils.py [hunk 3]
```diff
@@ -257,19 +264,39 @@ def train_loop(
...
+                                p for p in model.model.parameters() if p.requires_grad
+                            ]
+                            sci_grads = []
+                            for p in parameters:
+                                if (p.grad == None):
+                                    sci_grads.append(None)
+                                else:
+                                    param = p.grad.cpu().data
+                                    sci_grads.append(param)
+                            model.model.zero_grad()
```
Comment (mcgrath-k) [inline training_utils.py]: nit: code duplication from these lines between this and lines `310-326`. Not necessary, but if you have time deduplication is appreciated.

File: training_utils.py [hunk 4]
```diff
@@ -377,4 +431,69 @@ def train_loop(
...
+        total_train_sz = len(train_dataset)
+        if num_steps > max_steps:
+            break
+        with torch.enable_grad(), tqdm(total=total_train_sz) as progress_bar:
+            for curr_batch_dict in train_dataloader:
+                num_steps += 1
+                if num_steps > max_steps:
+                    break
+                if type(model) != torch.nn.Module:
+                    model.model.train()
```
Comment (mcgrath-k) [inline training_utils.py:482]: nit: This can probably be checked outside of the loop to simplify.

File: tree_projections/tree_projections_utils.py [hunk 5]
```diff
@@ -173,6 +177,12 @@ def relax_cond(mask, relax_mask, num_layers):
     batch_size = 256
     st = 0
 
+    if(diff):
```
Comment (mcgrath-k) [inline tree_projections/tree_projections_utils.py]: I didn't know this was allowed in python lol. Please add a space here. `if diff:`

File: training_utils.py [hunk 6]
```diff
@@ -377,4 +431,69 @@ def train_loop(
...
+
+def reg_loop(
+    args,
+    model,
+    train_dataset,
+    device,
+    tokenizer=None,
+    regularizer=None
+):
+    # Quick hack to get tree regularization scores for a saved model
```
Comment (mcgrath-k) [inline training_utils.py]: Does this comment apply to the entire `reg_loop`? The functionality of this is kinda unclear without a bit of time reading. A docstring for this function would be helpful! ~

File: regularizer_new.py [hunk 7]
```diff
@@ -38,16 +72,128 @@ def cache(self, input_str, parse_split=None):
...
+    def get_phrases(self, str):
+        # Sample phrases from given string
+        phrases = set()
+
+        # It is guaranteed that str is larger than phrase len
+        for iter in range(self.sample_num):
+            if (self.dataset in ["ds-addmult-mod10", "dyck"]):
+                # ensure phrases are the largest nested subtree
+                if (self.dataset == "dyck"):
+                    str = str.split(" ")
```
Comment (mcgrath-k) [inline regularizer_new.py]: Variable shadowing is best to avoid when possible.

File: regularizer_new.py [hunk 8]
```diff
@@ -103,83 +245,249 @@ def tokenizer_helper(s, add_special_tokens=True):
...
+                        if (st == 0):
+                            scores[str_idx][key] = 0
+                            continue
+                        else:
+                            inner_context_vectors = all_vector_idxs[st]
+                    else:
+                        inner_context_vectors = inner_context_vecs[idx][-1]
+                    scores[str_idx][key] = self.sim_metric(
+                        inner_context_vectors, fully_contextual_vectors
+                    )
```
Comment (mcgrath-k) [inline regularizer_new.py]: Dangerous to duplicate large sections like this. Large chunks like this can easily break in future commits / create unexpected errors.

Comment (mcgrath-k) [reply regularizer_new.py]: Please combine this and the section above from 271-295

File: regularizer_new.py [hunk 9]
```diff
@@ -103,83 +245,249 @@ def tokenizer_helper(s, add_special_tokens=True):
...
+                    s2 = recurse(score_chart, curr_str, best+1, en)
+                    
+                    if mean:
+                        norm_score = best_score - (tot_score/(en-st))
+                    else:
+                        k_rand = randint(st, en-1)
+                        norm_score = best_score - score_chart[(st,k_rand)] - score_chart[(k_rand+1,en)]
+
+                norm_score = min(norm_score, self.hinge_const)
                 return s1 + s2 + norm_score
```
Comment (mcgrath-k) [inline regularizer_new.py]: Generally, try to avoid nesting large functions like this. Keep nested functions to a minimum, or avoid all together. This can stay here for now, but this makes the readability / usage of a function much less clear for reviewers. Shouldn't problem for our codebase if we do not keep adding on top ~ advice for the future!

File: regularizer_new.py [hunk 10]
```diff
@@ -60,35 +206,31 @@ def tokenizer_helper(s, add_special_tokens=True):
             else:
                 return self.tokenizer(s)
```
Comment (mcgrath-k) [inline regularizer_new.py]: I know this is technically not part of this PR, but please put nested functions such as `tokenizer_helper` at the top of the function, it is much easier to read for others if this is not nested deeply inside a function.

**Review by mcgrath-k — CHANGES_REQUESTED** [review]
Looks good @ananjan-nandi-9!

I would highly recommend trying to spot largely nested blocks of `if / else` logic in massive functions, and also doing multiple things in functions. In `regularizer_new.py` the functions `build_scores` and `get_score` are too large. 

I understand there is a lot of algorithmic complexity, so there will ofc be a need for lots of conditional logic. Anywhere else where you can pull out logic into singular, modular functions that only tackle "one thing" ([Single Responsibility Principle](https://en.wikipedia.org/wiki/Single_responsibility_principle)) will always benefit. 

The obvious benefit is that others can read code easier, but I find that removing `if/else` nesting wherever possible allows for much quicker debug and error prevention. It also allows for software to expand without bloat, or change functionality easily. 

This type of design really helps out in industry, it allows your code to stay resilient when your team's priorities change, or if a quick new feature is needed. 

I can tell your code style is improving well (thanks for the extra work with the docstrings!)

Let me know if you have any questions!

File: data_utils/ds_addmult_mod10_helpers.py (hunk 1 above)
Comment (ananjan-nandi-9) [reply data_utils/ds_addmult_mod10_helpers.py]: That's odd, I think Github glitched with the merge. This is different on my local branch. Will resolve!

File: train_transformers.py (hunk 2 above)
Comment (ananjan-nandi-9) [reply train_transformers.py:214]: After the previous merge, all of these functions had been duplicated by Github. I just removed those duplicates in this MR.

File: training_utils.py (hunk 3 above)
Comment (ananjan-nandi-9) [reply training_utils.py]: Fixed, thanks for pointing out!

File: training_utils.py (hunk 4 above)
Comment (ananjan-nandi-9) [reply training_utils.py:482]: This is a holdout from Shikhar's code, some of the methods below access the Interface object, while others use the model itself.

File: tree_projections/tree_projections_utils.py (hunk 5 above)
Comment (ananjan-nandi-9) [reply tree_projections/tree_projections_utils.py]: Neither did I lol. Fixed.

File: training_utils.py (hunk 6 above)
Comment (ananjan-nandi-9) [reply training_utils.py]: Yup, it applies to the entire loop. Added docstring.

File: regularizer_new.py (hunk 7 above)
Comment (ananjan-nandi-9) [reply regularizer_new.py]: Changed, thanks!

File: regularizer_new.py (hunk 8 above)
Comment (ananjan-nandi-9) [reply regularizer_new.py]: Moved into functions.

File: regularizer_new.py (hunk 9 above)
Comment (ananjan-nandi-9) [reply regularizer_new.py]: I based my code structure for this on Shikhar's initial attempt at the regularizer. I get your point though, this should definitely have been a separate function. Moved out into a separate method.

File: regularizer_new.py (hunk 10 above)
Comment (ananjan-nandi-9) [reply regularizer_new.py]: Changed.

**Comment by ananjan-nandi-9** [comment]
Hi @mcgrath-k , implemented the changes as suggested by you. Also agree that get_score and build_scores are too large currently, but refactoring them will take some time. I think I will look into that after the paper is done. Thank you for the time spent on this!

File: training_utils.py (hunk 4 above)
Comment (mcgrath-k) [reply training_utils.py:482]: Ah gotcha, thanks for clarifying.

File: train_transformers.py (hunk 2 above)
Comment (mcgrath-k) [reply train_transformers.py:214]: Ah I did not see the duplicates, good catch.

File: training_utils.py (hunk 6 above)
Comment (mcgrath-k) [reply training_utils.py]: Thanks!

File: tree_projections/tree_projections_utils.py (hunk 5 above)
Comment (mcgrath-k) [reply tree_projections/tree_projections_utils.py]: XD

**Comment by mcgrath-k** [comment]
> Hi @mcgrath-k , implemented the changes as suggested by you. Also agree that get_score and build_scores are too large currently, but refactoring them will take some time. I think I will look into that after the paper is done. Thank you for the time spent on this!

When it comes to large functions like those, I agree, they can take a lot of time to refactor. When you plan on expanding on the functionality or radically modifying the function it is worth it sink the time investment into refactoring since it makes all future changes much easier. 

The best method is to try to design everything as modular right as you start coding, as long as it doesn't slow down your ideas by too much (I find modular design actually makes me code faster!). 

Everything looks great, thanks for combing through this. Go ahead with merge!

**Review by mcgrath-k — APPROVED** [review]
LGTM
